javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -80,121 +80,68 @@ object Arrays {
-      // pL is the start of the pivot block; move it into the middle if needed
-      if (pL != iK) {
-        a(pL) = a(iK)
-        a(iK) = pivot
-        pL = iK
-      }
-      // Elements equal to the pivot will be in range pL until pR
-      var pR = pL + 1
-      // Items known to be less than pivot are below iA (range i0 until iA)
-      var iA = i0
-      // Items known to be greater than pivot are at or above iB (range iB until iN)
-      var iB = iN
-      // Scan through everything in the buffer before the pivot(s)
-      while (pL - iA > 0) {
-        val current = a(iA)
-        ord.compare(current, pivot) match {
-          case 0 =>
-            // Swap current out with pivot block
-            a(iA) = a(pL - 1)
-            a(pL - 1) = current
-            pL -= 1
-          case x if x < 0 =>
-            // Already in place.  Just update indicies.
-            iA += 1
-          case _ if iB > pR =>
-            // Wrong side.  There's room on the other side, so swap
-            a(iA) = a(iB - 1)
-            a(iB - 1) = current
-            iB -= 1
-          case _ =>
-            // Wrong side and there is no room.  Swap by rotating pivot block.
-            a(iA) = a(pL - 1)
-            a(pL - 1) = a(pR - 1)
-            a(pR - 1) = current
-            pL -= 1
-            pR -= 1
-            iB -= 1
-        }
-      }
-      // Get anything remaining in buffer after the pivot(s)
-      while (iB - pR > 0) {
-        val current = a(iB - 1)
-        ord.compare(current, pivot) match {
-          case 0 =>
-            // Swap current out with pivot block
-            a(iB - 1) = a(pR)
-            a(pR) = current
-            pR += 1
-          case x if x > 0 =>
-            // Already in place.  Just update indices.
-            iB -= 1
-          case _ =>
-            // Wrong side and we already know there is no room.  Swap by rotating pivot block.
-            a(iB - 1) = a(pR)
-            a(pR) = a(pL)
-            a(pL) = current
-            iA += 1
-            pL += 1
-            pR += 1
-        }
-      }
-      // Use tail recursion on large half (Sedgewick's method) so we don't blow
-      // up the stack if pivots are poorly chosen
-      if (iA - i0 < iN - iB) {
-        quickSort(a, i0, iA)  // True recursion
-        quickSort(a, iB, iN)  // Should be tail recursion
+      insertionSort(a, i0, iN)
+    }
+  }
+
+  @inline
+  private def stableMerge[@specialized K](a: Array[K], b: Array[K], i0: Int,
+      iM: Int, iN: Int)(implicit ord: Ordering[K]): Unit = {
+    @tailrec def merge(k: Int, i: Int, j: Int): Unit = {
+      if (i < iM && (j >= iN || ord.lteq(a(i), a(j)))) {
+        b(k) = a(i)
+        if (k < iN)
+          merge(k + 1, i + 1, j)
       } else {
-        quickSort(a, iB, iN)  // True recursion
-        quickSort(a, i0, iA)  // Should be tail recursion
+        b(k) = a(j)
+        if (k < iN)
+          merge(k + 1, i, j + 1)
       }
     }
+    merge(i0, i0, iM)
```

Comment:
IMO this would be more readable with a good old while loop.

Also give meaningful names to the variables, e.g.,
* `k` -> `outIndex`
* `i` -> leftInIndex`
* `j` -> rightInIndex`